### PR TITLE
[Block Editor]: Add `flex-wrap` control to `flex` layout

### LINKED ISSUE #35525: Flex Layout: Allow control over flex-wrap
To build on the flex controls available to the Group block, we should have the ability to specify `flex-wrap`. This would allow us greater design flexibility over the responsive behavior of flex objects. 

Related: https://github.com/WordPress/gutenberg/issues/33687

### PR DESCRIPTION
**Pull request by ntsekouras**
Resolves: https://github.com/WordPress/gutenberg/issues/35525

This PR just adds a `flex-wrap` control in `flex` layout.
Screenshot 2021-11-01 at 4 04 08 PM

### CODE REVIEW

**Comment by kjellr** [comment]
This is great, and I can tell it will solve many of the responsive issues we've had with Twenty Twenty-Two header patterns: https://github.com/WordPress/twentytwentytwo/issues/89

wrap

Rather than "Flex Wrap", I wonder if we should just call this "Wrapping"?

File: packages/block-editor/src/layouts/flex.js
Comment (kjellr) [inline packages/block-editor/src/layouts/flex.js]:
```suggestion
		wrap: __( 'Items may wrap onto multiple lines.' ),
```

Just quick clarification that items won't _always_ wrap.

**Comment by shaunandrews** [comment]
Instead of using "flex" and "wrap" maybe we can explain this better with more straight-forward language. Also, as this is a binary (on/off) setting, maybe the segmented control is the wrong UI. Instead, maybe a switch will do what we need:

image

I'm not entirely sure what the default should be; Do you turn off "multiple lines" or turn on "multiple lines" — a lot of those decisions should be based around actual usage and general needs.

**Comment by Mamaduka** [comment]
I like @shaunandrews suggestion. I think that's the only information end-user need to use this feature.

The default value of `flex-wrap` is ~~`nowrap`~~ `wrap`, so the feature could be off by default.

Note to self: Browser defaults to `nowrap`, but our default is `wrap` 😄

File: lib/block-supports/layout.php
Comment (Mamaduka) [inline lib/block-supports/layout.php]:
```suggestion
		$flex_wrap_options = array( 'wrap', 'nowrap' );
		$flex_wrap         = ! empty( $layout['flexWrap'] ) && in_array( $layout['flexWrap'], $flex_wrap_options, true ) ?
			$layout['flexWrap'] :
			'wrap';
```

Key/value pairs are the same. So we can omit keys and use the options array for validation.

Comment (ntsekouras) [reply lib/block-supports/layout.php]: I actually used an object at first because in some cases (like justification) is better to map a key to some values, but in this case since it we will only have this two values, your suggestion makes more sense. I updated, thanks!

File: packages/block-editor/src/layouts/flex.js
Comment (ntsekouras) [reply packages/block-editor/src/layouts/flex.js]: The `allow` in the new label is similar to your suggestion, no?

**Review by Mamaduka — APPROVED** [review]
Thanks for working on this, Nik.

The change looks great codewise. I will leave the final ✅  to more design-savvy folks 😄 

https://user-images.githubusercontent.com/240569/139213536-1cb4e0f7-f227-468b-a5c0-5b5c437f4090.mp4

**Comment by jasmussen** [comment]
I wanted to leave a small note on this feature. 

On the one hand, it solves an immediate problem at hand, and gives power to the Row block and similar flex blocks, power that can solve responsive problems in an intrinsic way (see also #34641). 

At the same time, the feature itself is rather complicated for anyone unfamiliar with CSS. To an extent, the Row block can be thought of as a horizontal version of the group block: why wouldn't it always wrap? By virtue of being a property of the _flex_ type, this property also gets added to every flex block, including as demoed (thank you George!), Social Links. And therein you could get into trouble: disallow wrapping on it, and increase the gap. What happens? 

CSS and HTML has always allowed you to get into trouble. We can't entirely prevent that from happening in the block editor. But because of the _complexity_ of the wrapping feature, and because of the trouble it can get you in, before we land this we need to be absolutely sure that this is a feature we _want_ to have, and that we have considered all other alternatives before it. 

And if we do end up decicing: yes, this is a valuable feature, next we must manage the prominence of it. In context of the above, I would reduce the prominence as much as possible, due to it being in superuser territory. If we could change the Layout panel to a Tools Panel, then hide the wrapping feature in the ellipsis and never show it by default, we'd help balance the complexity, yet still let it be available for those who seek it out. (Balancing the prominence is also an argument for using the toggle 👍 👍)

Finally, there's something about the spacing and margin on this one that feels imbalanced:

Screenshot 2021-10-28 at 10 27 22

It's as if wrapping is related to justification when it's not. This is a bit of a general problem with toggles that we need to solve in light of #27331. It may just be to increase the margin?

**Comment by ntsekouras** [comment]
These are great concerns @jasmussen, thank you! I would love some folks with more experience in the css implications to share their thoughts.

Personally, after reading Joen's comments I might lean on landing this after the ToolsPanel change has landed 🤔

**Comment by kjellr** [comment]
> CSS and HTML has always allowed you to get into trouble. We can't entirely prevent that from happening in the block editor. But because of the complexity of the wrapping feature, and because of the trouble it can get you in, before we land this we need to be absolutely sure that this is a feature we want to have, and that we have considered all other alternatives before it.

It seems like the alternative here would be to intelligently handle wrapping by default. I find it a little hard to wrap my head around that though — it would take a lot of trial and error. 

I'm not sure if this is a helpful answer or not, but should this be the sort of feature we launch without any UI? Since as you note, folks can get into a lot of trouble with it, maybe it's only available for themes/patterns to use, until we come up with a safer/automated solution? One that will get folks into less trouble?

**Comment by shaunandrews** [comment]
I don't think its that complicated, nor do I think it will "get people into trouble" — at least not any more than any of the other settings on this block. (I'm staring directly at the bonkers "Inherit default layout" switch, which feels completely broken here...)

I think the langue could probably do a lot to help explain the setting. I went so far as to mockup a visual representation of the setting, but I'm not entirely sure its necessary.

image

To me, it's a switch that I'd toggle on and off a few times and see what happens. It's probably worth noting that this section will eventually make use of the Tools Panel, which means this switch will only be visible when needed.

**Review by kjellr — APPROVED** [review]
This will unblock us for Twenty Twenty-Two header issues, so I think we should we should add it for now. We already don't show any flex UI until that setting has been manually added to a block, so we have room to iterate on this and automate it moving forward.

**Comment by ntsekouras** [comment]
>We already don't show any flex UI until that setting has been manually added to a block

You mean `flex` in general? Because if you meant `flex-wrap`, the control right now is shown for blocks with `layout:flex` but it takes the default value(`wrap`) which is the current set value already.

**Comment by kjellr** [comment]
Yes, I mean flex in general. _All_ of the flex controls are somewhat hidden until they're needed, so I feel ok about adding this toggle there too.

**Pull request merged**
